### PR TITLE
Fix zfsctl_snapshot_{,un}mount() issues

### LINKED ISSUE #4377: grsec: cannot enter .zfs/snapshot/X
when trying to enter .zfs/snaphot/X i get:
`too many levels of symbolic links`

in dmesg there are alot of:

`grsec: denied exec of usermode helper binary /bin/sh located outside of permitted system paths
`
manual mount of the snapshots works

### PR DESCRIPTION
**Pull request by stiell**
Fix use after free in `zfsctl_snapshot_unmount()`. Use `/usr/bin/env` instead of `/bin/sh` to fix a shell code injection flaw and allow use with grsecurity.

Not yet tested on a kernel with grsecurity.

Fixes #4377. Supersedes #4568 and #4573.

### CODE REVIEW

**Comment by stiell** [comment]
Tested on Gentoo `hardened-sources-4.4.8-r1` (`grsecurity-3.1-4.4.8-201604252206`), automatic mounting and unmounting under `.zfs/snapshot` works.

**Comment by behlendorf** [comment]
@rlaager @tuxoko you both commented on the original PRs, could you also review this updated simpler version.

File: module/zfs/zfs_ctldir.c [hunk 1]
```diff
@@ -1032,10 +1032,10 @@ zfsctl_snapshot_unmount(char *snapname, int flags)
 
 	argv[2] = kmem_asprintf(SET_UNMOUNT_CMD,
 	    flags & MNT_FORCE ? "-f " : "", se->se_path);
-	zfsctl_snapshot_rele(se);
 	dprintf("unmount; path=%s\n", se->se_path);
 	error = call_usermodehelper(argv[0], argv, envp, UMH_WAIT_PROC);
 	strfree(argv[2]);
+	zfsctl_snapshot_rele(se);
```
Comment (tuxoko) [inline module/zfs/zfs_ctldir.c]: What's the point of this patch when your other patch negate the possibility of use-after-free?

Comment (tuxoko) [inline module/zfs/zfs_ctldir.c]: Oh sorry, the use-after-free is in `dprintf` not `strfree`. Then you should just move the `zfsctl_snapshot_rele` right after `dprintf`.

Comment (stiell) [inline module/zfs/zfs_ctldir.c]: The second patch also uses `se` in `argv`, so then `zfsctl_snapshot_rele` would have to be moved to after `call_usermodehelper` again in that patch.

Comment (tuxoko) [inline module/zfs/zfs_ctldir.c]: Right.

Comment (behlendorf) [inline module/zfs/zfs_ctldir.c]: @tuxoko can you approve the patch if you're happy with this.

**Pull request merged**